pkg/cloud-init/cloud-init.go

```diff
@@ -0,0 +1,376 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package cloudinit
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"strconv"
+
+	model "github.com/jeevatkm/go-model"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/logging"
+	"kubevirt.io/kubevirt/pkg/precond"
+)
+
+type IsoCreationFunc func(isoOutFile string, inFiles []string) error
+
+var cloudInitLocalDir = "/var/run/libvirt/cloud-init-dir"
+var cloudInitOwner = "qemu"
+var cloudInitIsoFunc = defaultIsoFunc
+
+const noCloudFile = "noCloud.iso"
+
+// Supported DataSources
+const (
+	dataSourceNoCloud = "noCloud"
+)
+
+func defaultIsoFunc(isoOutFile string, inFiles []string) error {
+
+	var args []string
+
+	args = append(args, "-output")
+	args = append(args, isoOutFile)
+	args = append(args, "-volid")
+	args = append(args, "cidata")
+	args = append(args, "-joliet")
+	args = append(args, "-rock")
+	args = append(args, inFiles...)
+
+	cmd := exec.Command("genisoimage", args...)
+	return cmd.Run()
+}
+
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	exists := false
+
+	if err == nil {
+		exists = true
+	} else if os.IsNotExist(err) {
+		err = nil
+	}
+	return exists, err
+}
+
+func md5CheckSum(filePath string) ([]byte, error) {
+	var result []byte
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return result, err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	_, err = io.Copy(hash, file)
+
+	if err != nil {
+		return result, err
+	}
+
+	result = hash.Sum(result)
+	return result, nil
+}
+
+func setFileOwnership(username string, file string) error {
+	usrObj, err := user.Lookup(username)
+	if err != nil {
+		return err
+	}
+
+	uid, err := strconv.Atoi(usrObj.Uid)
+	if err != nil {
+		return err
```

Comment:
It might be helpful throughout this module to also log this type of error before returning it.

Reply:
The error gets bubbled up through the config-disk package into virt-handler during the VM launch flow.  The error is in the reason field for "Synchronizing the VM failed." log message.

It wouldn't hurt to add a log message here in the package with a lesser log priority though.